### ISSUE TITLE
sprintf: fix number formatting quirks wrt float/int

### ISSUE THREAD

**Issue by srenatus** [open-policy-agent/opa#4610]
```
$ opa eval -fpretty 'sprintf("%f", [13])'
"%!f(int=13)"
```
This should output `"13"`. Using `%v` is a workaround, but won't allow us to control the number of digits after the dot, like `%.3f` would allow.

The problem is this code: https://github.com/open-policy-agent/opa/blob/v0.39.0/topdown/strings.go#L427-L436

☝️ We're trying to find the best format of the arg to pass along to the [golang stdlib Printf function](https://pkg.go.dev/fmt). The heuristic breaks down when you try to format a number that could be an int or a float internally.

Here's another example:

```
$ opa eval -fpretty --import future.keywords '{ sprintf("%f", [x]) | some x in {1, 1.1}}'
[
  "%!f(int=1)",
  "1.100000"
]
```

The desired output would be `[ "1", "1.100000"]`.

💭 One approach would be to examine the format string, and prepare the number argument accordingly: _If the n-th format specifier is of the float-y type, make the n-th arg a float; if it's int-y, make it an int._

**Comment by damienjburks** [comment]
I'd like to take a stab at completing this!

**Comment by srenatus** [comment]
Thanks! LMK if you need assistance or if something is unclear or if you'd like to discuss some approach... 😃

**Comment by damienjburks** [comment]
Question: @srenatus Let's say I set the format to %.3f in the example you've shown above. Would this be the expected output?

```
$ opa eval -fpretty --import future.keywords '{ sprintf("%.3f", [x]) | some x in {1, 1.1}}'
[
  "001",
  "1.100"
]
```

OR would you prefer to see the exact same with the int value?

```
$ opa eval -fpretty --import future.keywords '{ sprintf("%.3f", [x]) | some x in {1, 1.1}}'
[
  "1",
  "1.100"
]
```

**Comment by srenatus** [comment]
Hmm I think I'd go with `"1.000"`. Like here, https://go.dev/play/p/tNR1eM9TpSD

**Comment by hugorut** [comment]
Hi there, should we reopen this issue since the fix has been reverted? https://github.com/open-policy-agent/opa/pull/4794

Or is there another issue I should watch to track a fix for this behaviour?

**Comment by srenatus** [comment]
Yeah that other issue is what to watch now. We could reopen this one, anyways, you're right.

**Comment by srenatus** [comment]
@sspaink Do we still think this is a high priority given that we have string templates now? I think I'd rather get string templates to be the go-to solution, and sprintf can fade into insignificance 🌆

**Comment by anderseknert** [comment]
Interpolated strings still don't
> allow us to control the number of digits after the dot, like %.3f would allow.

But perhaps they should? 🙂 I too would rather invest in interpolation than `sprintf`, where possible.

**Comment by sspaink** [comment]
Investing into string interpolation seems like the right call, lowering the priority.

**Comment by srenatus** [comment]
> But perhaps they should? 🙂

Let's port some strconv builtins? I'd take a builtin variant of https://pkg.go.dev/strconv#FormatFloat over `sprintf`. Then you can put that into your interpolated string, `$"x: {strconv.format_float(1/3, 'f', -1, 32)}"` 💭

**Comment by anderseknert** [comment]
Heh, I guess I was more thinking of adding some `sprintf`-like behaviors to string interpolation. No idea what it'd look like, but if we followed `sprintf`, perhaps something like:

```rego
n := 1 / 3
s := $`{n}.3f`
# or
s := $`{n}{".3f"}`
```

or what do I know :) Adding a new built-in to fix an issue in an existing one feels a little... wrong, I think. But I guess the same can be said about fixing it via interpolation and not in `sprintf`. So thinking about it, unless we deprecate `sprintf`, let's perhaps fix this here first, and then make it easier/better to this by other means. A built-in function is a bug _in OPA_ after all.
